x/stake/keeper/val_state_change.go

```diff
@@ -130,6 +130,92 @@ func (k Keeper) ApplyAndReturnValidatorSetUpdates(ctx sdk.Context) (newVals []ty
 	return newVals, updates
 }
 
+// update staked tokens snapshots of all validators
+// elect topN validators according to the accumulated staked tokens over snapshotNum
+func (k Keeper) UpdateAndElectValidators(ctx sdk.Context) (newVals []types.Validator, updates []abci.ValidatorUpdate) {
+	snapshotNum := int(k.MaxStakeSnapshots(ctx))
+	var validators []types.Validator
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, ValidatorsKey)
+	defer iterator.Close()
+	var validator types.Validator
```

Comment:
Try init with `newVals = make([]types.Validator, 0, maxValidators)`

Reply:
`newVals` is just a slice of `validators`, do you mean init `validators` like this?
Since it includes all validators, we can't know an exact cap here in advance. The optimization may not be that effective.